fluent-langneg/src/locale.js

```diff
@@ -34,53 +34,63 @@ export default class Locale {
    * It also allows skipping the script section of the id, so `en-US` is
    * properly parsed as `en-*-US-*`.
    */
-  constructor(locale, range = false) {
+  constructor(locale) {
     const result = localeRe.exec(locale.replace(/_/g, "-"));
     if (!result) {
+      this.isWellFormed = false;
       return;
     }
 
-    const missing = range ? "*" : undefined;
+    const language = result[1] || undefined;
+    const script = result[2] || undefined;
+    const region = result[3] || undefined;
+    const variant = result[4] || undefined;
 
-    const language = result[1] || missing;
-    const script = result[2] || missing;
-    const region = result[3] || missing;
-    const variant = result[4] || missing;
-
-    this.language = language;
-    this.script = script;
-    this.region = region;
+    if (language) {
+      this.language = language.toLowerCase();
+    }
+    if (script) {
+      this.script = script[0].toUpperCase() + script.substr(1);
+    }
+    if (region) {
+      this.region = region.toUpperCase();
+    }
     this.variant = variant;
-    this.string = locale;
+    this.isWellFormed = true;
   }
 
   isEqual(locale) {
     return localeParts.every(part => this[part] === locale[part]);
   }
 
-  matches(locale) {
+  matches(locale, thisRange = false, otherRange = false) {
     return localeParts.every(part => {
-      return this[part] === "*" || locale[part] === "*" ||
-        (this[part] === undefined && locale[part] === undefined) ||
-        (this[part] !== undefined && locale[part] !== undefined &&
-        this[part].toLowerCase() === locale[part].toLowerCase());
+      return ((thisRange && this[part] === undefined) ||
+              (otherRange && locale[part] === undefined) ||
+              this[part] === locale[part]);
```

Comment:
Understanding this method is important for understanding the whole algorithm and I'm not fond of adding two boolean arguments to it. It's hard to follow what they mean when the method is called. In fact, I think I largely preferred the old code which used an explicit `*` to represent ranges. 

A few suggestions on how to improve this code. I'm not saying all of them need to be applied at the same time, but I'd like to see this method improved _somehow_.

- Remove the `thisRange` param as it is always `true` across all callsites, or start using it for exact matching.
- Use an `options` dictionary: `locale.matches(other, {thisRange: true, otherRange: true})`.
- Separate this code into multiple methods: `exactMatches`, `matchesThisRange`, `matchesRanges`.
- Add a flag to `Locale` called `undefinedAsRange` and a method called `toRange()` which returns a new `Locale` instance based on the one the method was called on. `match` would inspect the value of the flag for both `this` and `other` and act accordingly.

Now, I understand that this code is based on the existing C++/Rust code. With that in mind, I think this can be moved to a follow-up. Please file it if you agree.